File: vscode/src/services/BranchShareController.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";

const h = vi.hoisted(() => ({
	getShare: vi.fn(),
	putBranchShare: vi.fn(),
	removeShare: vi.fn(),
	revokeBranchShare: vi.fn(),
	updateLiveShare: vi.fn(),
}));

vi.mock("../../../cli/src/core/BranchShareStore.js", () => ({
	getShare: h.getShare,
	putBranchShare: h.putBranchShare,
	removeShare: h.removeShare,
}));
vi.mock("./JolliShareService.js", () => ({
	revokeBranchShare: h.revokeBranchShare,
	updateLiveShare: h.updateLiveShare,
}));

import { patchShareAudience, revokeShare } from "./BranchShareController.js";

beforeEach(() => {
	for (const fn of Object.values(h)) fn.mockReset();
});

describe("revokeShare", () => {
	it("revokes the server share and clears the record", async () => {
		h.getShare.mockResolvedValue({ shareId: "sh_1" });
		await revokeShare("/repo", "feature/x", "KEY");
		expect(h.revokeBranchShare).toHaveBeenCalledWith(undefined, "KEY", "sh_1");
		expect(h.removeShare).toHaveBeenCalledWith("/repo", "feature/x", undefined);
	});

	it("keys a commit share by its commit hash", async () => {
		const commitHash = "c".repeat(40);
		h.getShare.mockResolvedValue({ shareId: "sh_2" });
		await revokeShare("/repo", "feature/x", "KEY", commitHash);
		// "KEY" is not a decodable sk-jol- key, so the derived env key is undefined.
		expect(h.getShare).toHaveBeenCalledWith("/repo", "feature/x", undefined, commitHash);
		expect(h.removeShare).toHaveBeenCalledWith("/repo", "feature/x", commitHash);
	});

	it("skips the server call AND the local delete when no matching share (foreign/absent record preserved)", async () => {
		// getShare is env-scoped: a foreign-backend record reads as undefined. Deleting it
		// locally would orphan a still-live link, so the record is left on disk.
		h.getShare.mockResolvedValue(undefined);
		await revokeShare("/repo", "feature/x", "KEY");
		expect(h.revokeBranchShare).not.toHaveBeenCalled();
		expect(h.removeShare).not.toHaveBeenCalled();
	});
});

describe("patchShareAudience", () => {
	const ref = {
		kind: "branchCollection" as const,
		relativePath: "feature/x",
		covered: [{ commitHash: "a".repeat(40), summaryDocId: 11, attachmentDocIds: [12] }],
	};
	const MEMBER = {
		shareId: "sh_1",
		shareUrl: "https://acme.jolli.ai/share/branch/1/view",
		visibility: "org" as const,
		recipients: ["a@x.com"],
		ref,
		headCommitHash: "a".repeat(40),
		expiresAt: "2026-09-01T00:00:00.000Z",
		decisionCount: 4,
		titles: ["A"],
	};

	it("org→people: PATCHes visibility, keeps recipients and the member URL", async () => {
		h.getShare.mockResolvedValue(MEMBER);
		// The member URL is stable across the toggle — the server omits shareUrl.
		h.updateLiveShare.mockResolvedValue({ shareId: "sh_1", visibility: "people", recipients: ["a@x.com"] });

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "people" });

		expect(h.updateLiveShare).toHaveBeenCalledWith(undefined, "KEY", "sh_1", { visibility: "people" });
		expect(out).toMatchObject({ visibility: "people", recipients: ["a@x.com"], shareUrl: MEMBER.shareUrl, ref });
		const [, , record] = h.putBranchShare.mock.calls[0];
		expect(record).toMatchObject({ visibility: "people", shareUrl: MEMBER.shareUrl });
	});

	it("flip to public: re-issues the bearer URL and DROPS the recipients allowlist", async () => {
		h.getShare.mockResolvedValue(MEMBER);
		h.updateLiveShare.mockResolvedValue({
			shareId: "sh_1",
			visibility: "public",
			shareUrl: "https://acme.jolli.ai/share/tok_new?ref=1",
		});

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "public" });

		expect(out?.visibility).toBe("public");
		expect(out?.shareUrl).toBe("https://acme.jolli.ai/share/tok_new?ref=1");
		expect(out?.recipients).toBeUndefined();
		const [, , record] = h.putBranchShare.mock.calls[0];
		expect("recipients" in record).toBe(false);
	});

	it("reflects the requested tier even when the server echoes a stale visibility", async () => {
		// Repro of the UI-revert bug: flip org→public, but the PATCH response echoes
		// the old tier. The requested visibility must win so the record (and dropdown)
		// don't snap back to org.
		h.getShare.mockResolvedValue(MEMBER); // org
		h.updateLiveShare.mockResolvedValue({ shareId: "sh_1", visibility: "org" });

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "public" });

		expect(out?.visibility).toBe("public");
		const [, , record] = h.putBranchShare.mock.calls[0];
		expect(record.visibility).toBe("public");
		expect("recipients" in record).toBe(false); // public drops the allowlist
	});

	it("recipients change: sends the replacement allowlist and persists the server echo", async () => {
		const next = ["b@x.com", "c@x.com"];
		h.getShare.mockResolvedValue(MEMBER);
		h.updateLiveShare.mockResolvedValue({ shareId: "sh_1", visibility: "org", recipients: next });

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { recipients: next });

		expect(h.updateLiveShare).toHaveBeenCalledWith(undefined, "KEY", "sh_1", { recipients: next });
		expect(out?.recipients).toEqual(next);
		expect(out?.visibility).toBe("org");
	});

	it("preserves unchanged cached fields when the PATCH echoes only changed ones", async () => {
		h.getShare.mockResolvedValue(MEMBER);
		h.updateLiveShare.mockResolvedValue({ visibility: "people" });

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "people" });

		expect(out).toMatchObject({
			shareId: MEMBER.shareId,
			shareUrl: MEMBER.shareUrl,
			visibility: "people",
			recipients: MEMBER.recipients,
			expiresAt: MEMBER.expiresAt,
			ref,
			headCommitHash: MEMBER.headCommitHash,
			decisionCount: MEMBER.decisionCount,
			titles: MEMBER.titles,
		});
	});

	it("falls back to the patched visibility when the server echoes nothing", async () => {
		h.getShare.mockResolvedValue({ ...MEMBER, visibility: "org", recipients: undefined });
		h.updateLiveShare.mockResolvedValue({});

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { recipients: ["z@x.com"] });

		expect(out?.visibility).toBe("org"); // no patch.visibility, no echo → existing wins
		expect(out?.recipients).toEqual(["z@x.com"]); // from the patch
	});

	it("omits recipients entirely when none are known (member had none, patch/echo carry none)", async () => {
		h.getShare.mockResolvedValue({ ...MEMBER, visibility: "org", recipients: undefined });
		h.updateLiveShare.mockResolvedValue({});

		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "people" });

		expect(out?.visibility).toBe("people"); // patch.visibility wins over the missing echo
		expect(out?.recipients).toBeUndefined(); // no allowlist carried through
	});

	it("no-ops when there is no link", async () => {
		h.getShare.mockResolvedValue(undefined);
		const out = await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "people" });
		expect(out).toBeUndefined();
		expect(h.updateLiveShare).not.toHaveBeenCalled();
	});

	it("keys a commit share by its commit hash", async () => {
		const commitHash = "c".repeat(40);
		h.getShare.mockResolvedValue({ ...MEMBER, commitHash });
		h.updateLiveShare.mockResolvedValue({ shareId: "sh_1", visibility: "people" });
		await patchShareAudience("/repo", "feature/x", "KEY", { visibility: "people" }, commitHash);
		expect(h.getShare).toHaveBeenCalledWith("/repo", "feature/x", undefined, commitHash);
		expect(h.putBranchShare).toHaveBeenCalledWith("/repo", "feature/x", expect.any(Object), commitHash);
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/services/BranchShareControllerTest.kt
package ai.jolli.jollimemory.services

import ai.jolli.jollimemory.core.BranchShareStore
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockkObject
import io.mockk.slot
import io.mockk.unmockkAll
import io.mockk.verify
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class BranchShareControllerTest {

    private val root = "/repo"
    private val branch = "feature/x"
    private val apiKey = "sk-jol-test"

    private fun record(
        id: String = "42",
        visibility: String = "public",
        recipients: List<String>? = null,
    ) = BranchShareStore.BranchShareRecord(
        shareId = id,
        shareUrl = "https://acme.jolli.ai/s/tok",
        visibility = visibility,
        recipients = recipients,
        expiresAt = "2026-12-31T00:00:00Z",
        decisionCount = 2,
    )

    @BeforeEach
    fun setUp() {
        mockkObject(BranchShareStore)
        mockkObject(JolliApiClient)
        every { BranchShareStore.putBranchShare(any(), any(), any(), any()) } returns Unit
        every { BranchShareStore.removeShare(any(), any(), any()) } returns Unit
    }

    @AfterEach
    fun tearDown() = unmockkAll()

    @Test
    fun `revoke calls the server then clears the local record`() {
        every { BranchShareStore.getShare(root, branch, null) } returns record(id = "99")
        every { JolliApiClient.revokeShare(any(), any(), any()) } returns Unit

        BranchShareController.revokeShare(root, branch, apiKey)

        verify { JolliApiClient.revokeShare(null, apiKey, "99") }
        verify { BranchShareStore.removeShare(root, branch, null) }
    }

    @Test
    fun `revoke with no existing share only clears locally`() {
        every { BranchShareStore.getShare(root, branch, null) } returns null
        BranchShareController.revokeShare(root, branch, apiKey)
        verify(exactly = 0) { JolliApiClient.revokeShare(any(), any(), any()) }
        verify { BranchShareStore.removeShare(root, branch, null) }
    }

    @Test
    fun `patchShareAudience to people persists the recipients allowlist`() {
        every { BranchShareStore.getShare(root, branch, null) } returns record(visibility = "public")
        every { JolliApiClient.updateLiveShare(any(), any(), any(), any()) } returns
            JolliApiClient.LiveShareUpdateResult(visibility = "people", recipients = listOf("a@x.com"))
        val stored = slot<BranchShareStore.BranchShareRecord>()
        every { BranchShareStore.putBranchShare(root, branch, capture(stored), null) } returns Unit

        val out = BranchShareController.patchShareAudience(
            root, branch, apiKey,
            BranchShareController.ShareAudiencePatch(visibility = "people", recipients = listOf("a@x.com")),
        )

        out!!.visibility shouldBe "people"
        stored.captured.recipients shouldBe listOf("a@x.com")
    }

    @Test
    fun `patchShareAudience to public drops the recipients allowlist`() {
        every { BranchShareStore.getShare(root, branch, null) } returns record(visibility = "people", recipients = listOf("a@x.com"))
        every { JolliApiClient.updateLiveShare(any(), any(), any(), any()) } returns
            JolliApiClient.LiveShareUpdateResult(visibility = "public")
        val stored = slot<BranchShareStore.BranchShareRecord>()
        every { BranchShareStore.putBranchShare(root, branch, capture(stored), null) } returns Unit

        BranchShareController.patchShareAudience(root, branch, apiKey, BranchShareController.ShareAudiencePatch(visibility = "public"))

        stored.captured.visibility shouldBe "public"
        stored.captured.recipients.shouldBeNull()
    }

    @Test
    fun `patchShareAudience with no share returns null`() {
        every { BranchShareStore.getShare(root, branch, null) } returns null
        BranchShareController.patchShareAudience(root, branch, apiKey, BranchShareController.ShareAudiencePatch(visibility = "org")).shouldBeNull()
    }
}
